File: Min_Cost_to_Connect_All_Points.js
/**
 * @param {number[][]} points
 * @return {number}
 */
var minCostConnectPoints = function(points) {
    const n = points.length;
    let minCost = 0;
    const visited = new Array(n).fill(false);
    const distances = new Array(n).fill(Infinity);

    distances[0] = 0;  // Start from the first point

    for (let i = 0; i < n; i++) {
        let u = getMinDistanceVertex(visited, distances);
        visited[u] = true;
        minCost += distances[u];

        for (let v = 0; v < n; v++) {
            if (!visited[v]) {
                const dist = Math.abs(points[u][0] - points[v][0]) + Math.abs(points[u][1] - points[v][1]);
                distances[v] = Math.min(distances[v], dist);
            }
        }
    }

    return minCost;    
};

var getMinDistanceVertex = function(visited, distances) {
    let minDist = Infinity;
    let minVertex = -1;

    for (let i = 0; i < distances.length; i++) {
        if (!visited[i] && distances[i] < minDist) {
            minDist = distances[i];
            minVertex = i;
        }
    }

    return minVertex;
};

File: Min_Cost_to_Connect_All_Points.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        min_cost = 0 # return value
        visited = [False] * n # to track visited points
        pq = [(0, 0)] # to store edge(cost, vertex)
        cache = {0: 0}

        while pq:
            # Get the edge with the minimum cost from the pq
            cost, u = heapq.heappop(pq)

            # if the vertex has already been visited, skip it
            if visited[u]:
                continue
            
            visited[u] = True
            min_cost += cost

            # Iterate through all the other vertices
            for v in range(n):
                if not visited[v]:
                    # calculate the distance between current point(u) and neighbor point(v)
                    dist = abs(points[u][0] - points[v][0]) + abs(points[u][1] - points[v][1])
                    if dist < cache.get(v, dist + 1):
                        cache[v] = dist
                        heapq.heappush(pq, (dist, v))
        
        return min_cost
